social_core/backends/id4me.py

```diff
@@ -0,0 +1,270 @@
+        for key in self.get_agent_keys():
+            header = jwt.get_unverified_header(id_token)
+            if header['kid'] == key['kid']:
+                if 'alg' not in key:
+                    key['alg'] = 'RS256' if key['kty'] == 'RSA' else 'ES256'
+                return key
+
+    def auth_complete(self, *args, **kwargs):
+        self.validate_state()
+        identity = self.strategy.session_get(self.name + '_identity')
+        openid_configuration = self.get_identity_record(identity)
+        if 'clp' not in openid_configuration:
+            raise AuthUnreachableProvider(self)
+        self.strategy.session_set(self.name + '_agent', openid_configuration['clp'])
+        return super().auth_complete(*args, **kwargs)
+
+    def auth_params(self, state=None):
+        client_id, client_secret = self.get_key_and_secret()
+        params = {
+            'client_id': client_id,
+            'redirect_uri': self.get_redirect_uri(state)
+        }
+        if self.STATE_PARAMETER and state:
+            params['state'] = state
+        if self.RESPONSE_TYPE:
+            params['response_type'] = self.RESPONSE_TYPE
+
+        params.update({
+            'client_id': client_id,
+            'redirect_uri': self.get_redirect_uri(state),
+            'login_hint': self.strategy.session_get(self.name + '_identity')
+        })
+        if self.STATE_PARAMETER and state:
+            params['state'] = state
+        if self.RESPONSE_TYPE:
+            params['response_type'] = self.RESPONSE_TYPE
+        return params
+
+    def auth_url(self):
+        if not self.data.get('identity', ''):
+            raise AuthMissingParameter(self, 'identity')
+        identity = self.data.get('identity')
+        if not is_valid_domain(identity):
+            raise AuthForbidden(self)
+        openid_configuration = self.get_identity_record(identity)
+        if 'iss' not in openid_configuration:
+            raise AuthUnreachableProvider(self)
+        self.strategy.session_set(self.name + '_authority', openid_configuration['iss'])
+        self.strategy.session_set(self.name + '_identity', identity)
+        return super(ID4meBackend, self).auth_url()
+
+    def auth_complete_params(self, state=None):
+        data = {
+            'grant_type': 'authorization_code',
+            'code': self.data.get('code', ''),
+            'redirect_uri': self.get_redirect_uri()
+        }
+        return '&'.join(["{}={}".format(key, value) for key, value in data.items()])
+
+    def auth_complete_credentials(self):
+        return self.get_key_and_secret()
+
+    def validate_claims(self, id_token):
+        utc_timestamp = timegm(datetime.datetime.utcnow().utctimetuple())
+
+        if 'nbf' in id_token and utc_timestamp < id_token['nbf']:
+            raise AuthTokenError(self, 'Incorrect id_token: nbf')
+
+        # Verify the token was issued in the last 10 minutes
+        iat_leeway = self.setting('ID_TOKEN_MAX_AGE', self.ID_TOKEN_MAX_AGE)
```

Comment:
Why this arbitrary check for 10 minutes? There is `exp` claim which shall be respected.